[shard] ShardedTensor Interface

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74695



ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

Commit:
[WIP] ShardedTensor Interface

WIP ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

[ghstack-poisoned]

Commit:
Update on "[WIP] ShardedTensor Interface"


WIP ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

[ghstack-poisoned]

Commit:
Update on "[WIP] ShardedTensor Interface"


WIP ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

[ghstack-poisoned]

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -62,7 +61,53 @@ def _register_remote_shards(sharded_tensor_id: int, rrefs: List[rpc.RRef[Shard]]
+    def sharding_spec(self) -> shard_spec.ShardingSpec:
+        raise NotImplementedError()
+
+    def local_tensor(self) -> torch.Tensor:
+        raise NotImplementedError()
```
Maybe we remove `local_tensor` from the interface since it is mostly a specialization of `local_shards`.

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

wanchaol commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1026,9 +1027,7 @@ def test_sharded_tensor_sizes(self):
         # Test with invalid input
         st = sharded_tensor.empty(spec, (10, 20), init_rrefs=True)
-        with self.assertRaisesRegex(ValueError, 'must be within the range of tensor dimensions \\[0, 2\\)'):
-            st.size(-1)
```
added

wanchaol commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -448,7 +449,7 @@ def test_sharded_tensor_metadata(self):
         # the global metadata without changing the underlying shard's properties
-        with self.assertRaisesRegex(AttributeError, "can't set attribute"):
+        with self.assertRaisesRegex(RuntimeError, "torch function '__set__'"):
             st.requires_grad = True
```
created in https://github.com/pytorch/pytorch/issues/75272

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -685,6 +741,10 @@ def local_tensor(self) -> torch.Tensor:
     @classmethod
     def __torch_function__(cls, func, types, args=(), kwargs=None):
+        # Access attributes as is usually done on Tensors
+        if func.__name__ in ["__get__", "size"]:
+            return super().__torch_function__(func, types, args, kwargs)
```
So I tried it by making it a separate op, the biggest issue here is that if we make it a separate op using `sharded_op_impl`, we couldn't call `super()` because the cls instance is not passed to the separate op function, we can still call `super()` inside a classmethod and it will point to the correct instance. So I guess we can keep these two here until we find a better way.

pritamdamania87 replied:
Do we need super() here? Can't we just do this in the `sharded_op_impl`:

```
with torch._C.DisableTorchFunction():
  func(*args, **kwargs)
```

wanchaol replied:
yeah that's sth we can do, I just tried this approach and the difference is that we need to register all the property getters separately (i.e. `Tensor.shape.__get__`, `Tensor.dtype.__get__`) instead of a single `get` check here. But I feel that might be cleaner on what we are trying to implement, I'll update.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -352,18 +392,26 @@ def _init_from_local_shards(
         sharded_tensor._prepare_init(process_group=process_group, init_rrefs=init_rrefs)
 
         # add to metadata and local_shards
         sharded_tensor._metadata = global_sharded_tensor_metadata
         sharded_tensor._local_shards = local_shards
```
> Maybe we don't need to store this field statically, the ground truth metadata can be located inside the List[Shard], when user call ShardedTensor.metdata(), we can dynamically do an all_gather and return a ShardedTensorMetadata. What do you think about this?

A bit worried about this due to the additional all_gather call since I'm assuming there might be already usages of `.metadata()`. Btw, can't we move `_metadata` to the base class? We should be able to create the metadata from the sharding_spec and sizes/dtype etc. right?

Maybe we can do the reverse and store ground truth as ShardedTensorMetadata and dynamically build `List[Shard]` when users call `local_shards()` since that can be built without any comms.

wanchaol replied:
Agreed that can move `_metadata` to the base class, the trickly part is that `ShardedTensorMetadata` contains `ShardMetadata`, which is something that extension user might not want to use (i.e. they might want their own custom shardmetadata). In this case they will have two source of truths on sharding layout (1. our ShardMetadata layout built by `sharding_spec.build_metadata` 2. their custom shardmetadata layout). I guess they can silently ignore the first one and use their own layout description in this case?

pritamdamania87 replied:
Not sure I followed this completely, the way the extension interface works is that `sharding_spec.build_metadata` returns `ShardedTensorMetadata`, so the extension has to implement a way to return this metadata to us. So we can always call that method for metadata.

Now, the user can do their own thing, but then thats on their plate to deal with two source of truths. Ideally, they should just use `ShardedTensorMetadata` to avoid this.

fduwjj replied:
Yep, we call `.metadata()` in the sharded ops a lot, for example, chunk. So it will be nice that if we can avoid the communication here.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/interface.py:
```diff
@@ -0,0 +1,58 @@
+        raise RuntimeError(f"A {cls.__name__} object is being used from c++ while calling {func.__module__}.{func.__name__} "
+                           "but the there is no custom __torch_dispatch__ implementation for it.")
+
+    @abc.abstractmethod
+    def sharding_spec(self) -> shard_spec.ShardingSpec:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def local_shards(self) -> List[Shard]:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def reshard(self, resharding_spec: shard_spec.ShardingSpec) -> "ShardedTensorInterface":
+        raise NotImplementedError()
```
We need docs for these methods?

yeounoh commented on torch/distributed/_shard/sharded_tensor/interface.py:
```diff
@@ -0,0 +1,58 @@
+            pin_memory=pin_memory,
+            requires_grad=requires_grad
+        )
+        return r
```
`r` is referencing to the unpartitioned tensor, that would be sharded according to the spec? Also, would it be possible to pass a tensor to the constructor and create a ShardedTensorInterface -- or it's not needed?

yeounoh replied:
If `r` doesn't have any memory attached (which is desirable), do we implicitly collect and show the unpartitioned, `r` in this case, when accessed? Also, is there a variable pointing to the individual shards/handles, or it's up to the child implementation to deifne?

wanchaol replied:
r is merely a wrapper tensor with no memory attached as you can see, you can define both global tensor or local shards in the subclass. When accessing the instance of the ShardedTensor interface in python, it will be directed to `__torch_function__` (for python only ops like `__get__`, and to `__torch_dispatch__`, where you can control how the access should behave(i.e. return attributes of the full tensors, operating on individual shards, etc.)

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/interface.py:
```diff
@@ -11,19 +11,29 @@
+    which represents Tensors that are sharded across multiple devices and
+    multiple processes.
+
+    :class:`ShardedTensor` is a concrete implementation of this interface
+    and attached local shards with real data/memory.
```
Typically an interface shouldn't refer to its implementations in docs or otherwise. We can add such docs in ShardedTensor itself.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/interface.py:
```diff
@@ -44,15 +54,41 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
         raise NotImplementedError()
 
     def __hash__(self):
+        """
+        By default implementation to define behavior when `hash()` is called on an instance
+        of class.
+
+        Subclass of :class:`ShardedTensorInterface` could override this method if it choose
+        to return a different result.
+        """
```
Oh I guess we don't need docs for this method since it is not a public method rather a private magic method.

fduwjj commented on test/distributed/_shard/sharded_tensor/ops/test_linear.py:
```diff
@@ -190,7 +190,7 @@ def test_sharded_linear_errors(self):
         for spec in generate_chunk_sharding_specs_for_test(0):
             fc1 = torch.nn.Linear(10, 10).cuda(self.rank)
             shard_parameter(fc1, "bias", spec)
-            with self.assertRaisesRegex(TypeError, 'bias needs to be torch.Tensor'):
```
This is not actually what we want to test here. Maybe instead of changing the error msg, can you add one more line of shard_parameter(fc1, "weight", spec) so that the error msg will be triggered?

fduwjj reviewed (commented):
Thanks for doing this and this makes the op enabling a lot easier and cleaner!

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/linear.py:
```diff
@@ -125,7 +125,7 @@ def sharded_linear(types, args, kwargs, pg):
 def _validate_linear_op_param(args, kwargs):
     """
-    Validate input params of sharded embedding op.
+    Validate input params of sharded linear op.
```
Opps, nice catch!

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/linear.py:
```diff
@@ -141,13 +141,13 @@ def _validate_linear_op_param(args, kwargs):
     if not isinstance(input, torch.Tensor) and not isinstance(input, ShardedTensor):
         raise TypeError("input needs to be either torch.Tensor or ShardedTensor")
-    if not isinstance(bias, torch.Tensor):
+    if isinstance(bias, ShardedTensor):
```
Nit: this may not needed right?

wanchaol replied:
Yeah I think we can use `issubclass` instead, let me update it.

wanchaol replied:
Actually this is tricker, because `bias` can either be a `torch.Tensor` or `torch.nn.Parameter`. so a way to check an object is torch.Tensor and not its subclass will fail in the case when `bias` is a `torch.nn.Parameter`. 

So a robust way might be `type(bias) != torch.Tensor and type(bias) != nn.Parameter`, which is a bit nasty but I guess we could do that for now

fduwjj reviewed (approved):
Overall it looks good to me. Can we also ensure all the mypy warnings are taken care of?

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -408,18 +388,25 @@ def _init_from_local_shards(
-        sharded_tensor._sharding_spec = shard_spec._infer_sharding_spec_from_shards_metadata(
-            global_sharded_tensor_metadata.shards_metadata
-        )
+        sharded_tensor._sharding_spec = spec
```
Why do we need to do this? Aren't we already passing in the spec above?

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]

Commit:
Update on "[shard] ShardedTensor Interface"




ShardedTensor Interface:
1. Make a ShardedTensorInterface class that is a subclass of torch.Tensor, define basic APIs in ShardedTensorInterface, ShardedTensor is now a subclass of it.
2. By default disable `__torch_dispatch__` in the ShardedTensorInterface, ShardedTensor will use `__torch_function__` for now, subclasses of ShardedTensorInterface can use `__torch_dispatch__` by overriding it
3. remove attribute functions in ShardedTensor and handled them in `__torch_function__`

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123200/)!

Differential Revision: [D35123200](https://our.internmc.facebook.com/intern/diff/D35123200)

[ghstack-poisoned]
